### ISSUE TITLE
Performance on Real-Time Scenarios 

### ISSUE THREAD

**Issue by etarakci-hvl** [rohanchandra30/Spectral-Trajectory-and-Behavior-Prediction#32]
Hello, 

This is not an explicit issue, but I have been wondering how can one test your implementation in a real-time scenario? 
As far as I can see, output regarding the evaluation is given for datasets. Given that I have a memory of last N steps for M agents, what would you suggest in order to compute the prediction output for those M agents, without loss of generality? 

P.S. I have been trying to check your architecture's capabilities in a simulation environment.

**Comment by rohanchandra30** [comment]
By setting the code in EVAL mode (TRAIN=False), you would need to modify the dataloader to read your custom simulation data, and not extract a single batch from the dataset as it does now.

**Comment by etarakci-hvl** [comment]
Thank you for your answer. Let me elaborate my question.

Just to make sure I understand it correctly, given that **generate** function's _out tensor_ has this format: [Batch Size, Prediction Length, x/y] : (128,10,2). 
1) Do we presume that each bit in a batch represent a frame? 


For the prediction part, there is a function called generate(). I pass 
_train, encoder_stream1, decoder_stream1_ into it, let's keep that in mind. When you employ _compute_accuracy_stream1_, you calculate MSE, and in MSE function, there is a part I could not interpret.  y_gt (or label) values 
```
x = np.array(y_gt[:,:, 0])
x = (x-np.mean(x))/x.max() 
```
are processed to be compared with muX and muY. 

Here is the main question:
2) Given that the outputs below, how may I receive **trajectory prediction** results? Regarding the MSE function, as I mentioned, **label** values are processed to be compared with **pred** results.  **Processed label** and **pred** values are in their own format, yet **train** data is different. how so?
3) What sort of mechanism should be applied to induce future trajectories in a given format in **train**? Since, instead of giving direct trajectory outputs, implementation provides indirect comparison between **processed label** and **pred** and then provide ADE and FDE. Do I misinterpret or miss something?

Outputs:
print(train)
         [[3.2800, 1.8670],
         [3.3700, 2.0390],
         [4.0630, 2.0180],
         [4.0880, 1.9890],
         [4.1050, 2.0400],
         [3.5410, 2.0420]],

        [[3.3700, 2.0390],
         [4.0630, 2.0180],
         [4.0880, 1.9890],
         [4.1050, 2.0400],
         [3.5410, 2.0420],
         [3.4000, 2.0740]]

print(label)
         [3.4000, 2.0740],
         [3.5530, 2.0760],
         [3.9860, 2.0840],
         [4.1140, 1.9810],
         [3.7430, 1.7980],
         [4.1120, 1.8830],
         [3.2690, 1.8370],
         [3.2660, 1.8030],
         [3.4920, 1.8500],
         [3.2500, 1.8340]],

        [[3.5530, 2.0760],
         [3.9860, 2.0840],
         [4.1140, 1.9810],
         [3.7430, 1.7980],
         [4.1120, 1.8830],
         [3.2690, 1.8370],
         [3.2660, 1.8030],
         [3.4920, 1.8500],
         [3.2500, 1.8340],
         [3.4210, 2.0380]

print(pred)

         [[ 8.68478767e-02 -6.53539543e-01]
          [-4.29357955e-01 -6.09016797e-01]
          [-5.25244192e-02 -4.09469750e-05]
          [-9.53070995e-02  1.44321982e-01]
          [ 4.08940034e-01  5.74814333e-01]
          [-5.89859222e-02  1.76024020e-01]
          [-3.99820601e-01 -6.03259150e-01]
          [ 2.92644500e-01  2.29295889e-02]
          [ 2.86104653e-01  1.24864849e-01]
          [ 7.53063674e-02 -1.30372165e-01]]

         [[-2.39893287e-01 -3.72074273e-01]
          [ 9.62385531e-03  2.50858267e-01]
          [ 1.66076979e-02  2.81630824e-01]
          [-2.33874691e-01 -1.55309067e-01]
          [-3.30491792e-01 -3.75691175e-01]
          [-1.75945318e-01 -4.36439632e-02]
          [ 2.52300788e-01 -1.09749243e-01]
          [ 2.20098780e-01  1.49811623e-01]
          [-2.15301983e-01 -2.87120526e-01]
          [-1.68743163e-01 -7.84256442e-02]]

**Comment by etarakci-hvl** [comment]
To be precise, you apply following for the Ground Truth X and Y: 
    ```
    x = np.array(y_gt[:,:, 0])
    x = (x-np.mean(x))/x.max()
    y = np.array(y_gt[:,:, 1])
    y = (y-np.mean(y))/y.max()
    ```
In order to align with unprocessed Label outputs what do you think about the following expression? 
        
        ```
        pred = generate(train, encoder_stream1, decoder_stream1)
        
        base_x = np.array(train[:,:, 0])
        future_x = (pred_x + np.mean(base_x))

        base_y = np.array(train[:,:, 1])
        future_y = (pred_y + np.mean(base_y))
        ```

**Comment by rohanchandra30** [comment]
Why do you do the the following: base_x = (base_x[:,-1] + base_x[:,-2])/2  ?? and same for y?

**Comment by etarakci-hvl** [comment]
For that part, experimentally, I received lesser errors when used latest observations. When trying to map predicted results over real x, y values I observed that if full observation set is used, the ADE and FDE are increased. Thus, last set of x, y values are _relatively_ provide better base for predicted values to map onto.

**Comment by etarakci-hvl** [comment]
All in all, basic question is how would you acquire denormalized data? 
How would you map prediction results to the coordinate system of the given dataset?

**Comment by rohanchandra30** [comment]
As of now, the only way to produce original coordinates is to remove the normalization applied in the code.

**Comment by abojja9** [comment]
Hi,

I also got stuck in the same problem. I'm using the following transformation to map the predictions results to the coordinate system.

```python
    # Label
    x_ = np.array(label[:, 0], copy=True)
    y_ = np.array(label[:, 1], copy=True)
    # Prediction
    pred_x = np.array(pred[:, 0], copy=True)
    pred_y = np.array(pred[:, 1], copy=True)
    x_pred = np.mean(x_) + (pred_x - np.mean(pred_x)) * (np.std(x_)/np.std(pred_x))
    y_pred = np.mean(y_) + (pred_y - np.mean(pred_y)) * (np.std(y_)/np.std(pred_y))
```
I'm trying to visualise the prediction results, but they don't make any sense although MSE seems good. So, my transformation function seems wrong.
![viz](https://user-images.githubusercontent.com/31553857/93232284-b931aa80-f747-11ea-897c-7fee8ebfd175.png)

@rohanchandra30 could you please point where to remove normalization in the code?

@etarakci-hvl were you able to solve the problem? 

Thanks!
